apiel/src/parse/val.rs

```diff
@@ -0,0 +1,257 @@
+use conv::ConvUtil;
+use eyre::Result;
+use num_traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedNeg, CheckedSub};
+
+#[derive(Debug, Clone, Copy)]
+pub enum Val {
+    Float(f64),
+    Integer(i64),
+}
+
+impl Val {
+    fn promote_pair(a: &Val, b: &Val) -> (Val, Val) {
+        match (a, b) {
+            (Val::Integer(i), Val::Integer(j)) => (Val::Integer(*i), Val::Integer(*j)),
+            (Val::Float(f), Val::Float(g)) => (Val::Float(*f), Val::Float(*g)),
+            (Val::Integer(i), Val::Float(f)) => (Val::Float(*i as f64), Val::Float(*f)),
+            (Val::Float(f), Val::Integer(i)) => (Val::Float(*f), Val::Float(*i as f64)),
+        }
+    }
+}
+
+// #[derive(Debug, Clone)]
+// enum ValArray {
+//     Single(Val),
+//     Array(Vec<ValArray>),
+// }
+
+// impl From<Val> for ValArray {
+//     fn from(val: Val) -> Self {
+//         Self::Single(val)
+//     }
+// }
+
+impl TryFrom<Val> for usize {
+    type Error = &'static str;
+
+    fn try_from(value: Val) -> Result<Self, Self::Error> {
+        match value {
+            Val::Integer(val) => val
+                .try_into()
+                .map_err(|_| "Failed to convert i64 into usize"),
+            Val::Float(val) => {
+                if val.fract() == 0.0 && val >= 0.0 {
+                    val.approx_as::<usize>()
+                        .map_err(|_| "Failed to convert f64 into usize")
+                } else {
+                    Err("Float is not a whole number or is negative")
+                }
+            }
+        }
+    }
+}
+
+impl From<Val> for f64 {
+    fn from(value: Val) -> Self {
+        match value {
+            Val::Integer(val) => val as f64,
+            Val::Float(val) => val,
+        }
+    }
+}
+
+impl PartialEq for Val {
+    fn eq(&self, other: &Val) -> bool {
+        match (self, other) {
+            (Val::Integer(i), Val::Integer(j)) => i == j,
+            (Val::Float(f), Val::Float(g)) => f == g,
+            (Val::Integer(i), Val::Float(f)) => *i as f64 == *f,
+            (Val::Float(f), Val::Integer(i)) => *f == *i as f64,
+        }
+    }
+}
+
+impl Eq for Val {}

```

Comment:
Wow, magic. Not sure what this does, but consider using derive macro

Reply:
I can't derive Eq for floats :)